src/chainstate/burn/db/burndb.rs

```diff
@@ -0,0 +1,1767 @@
+        }
+        Ok(db)
+    }
+
+    /// Open a burn database in memory (used for testing)
+    pub fn connect_memory(first_block_height: u64, first_burn_hash: &BurnchainHeaderHash) -> Result<BurnDB<A, K>, db_error> {
+        let conn = Connection::open_in_memory()
+            .map_err(|e| db_error::SqliteError(e))?;
+
+        let mut db = BurnDB {
+            conn: conn,
+            readwrite: true,
+            tx_active: false,
+            first_block_height: first_block_height,
+            first_burn_header_hash: first_burn_hash.clone(),
+            
+            _phantom_a: PhantomData,
+            _phantom_k: PhantomData
+        };
+
+        db.instantiate()?;
+        Ok(db)
+    }
+
+    pub fn conn(&self) -> &Connection {
+        &self.conn
+    }
+
+    /// Get the previous consensus hashes that must be hashed to find
+    /// the *next* consensus hash at a particular block.
+    /// The resulting list will include the consensus hash at block_height.
+    fn get_prev_consensus_hashes(conn: &Connection, block_height: u64, first_block_height: u64) -> Result<Vec<ConsensusHash>, db_error> {
+        let mut i = 0;
+        let mut prev_chs = vec![];
+        while block_height - (((1 as u64) << i) - 1) >= first_block_height {
+            let prev_block : u64 = block_height - (((1 as u64) << i) - 1);
+            let prev_ch_opt = BurnDB::<A, K>::get_consensus_at(conn, prev_block)?;
+            match prev_ch_opt {
+                Some(prev_ch) => {
+                    debug!("Consensus at {}: {}", prev_block, &prev_ch.to_hex());
+                    prev_chs.push(prev_ch.clone());
+                    i += 1;
+
+                    if block_height < (((1 as u64) << i) - 1) {
+                        break;
+                    }
+                }
+                None => {
+                    error!("Failed to read consensus hash for block height {}", prev_block);
+                    return Err(db_error::Corruption);
+                }
+            };
+        }
+        Ok(prev_chs)
+    }
+
+    /// Make a new consensus hash, given the ops hash 
+    fn make_consensus_hash(conn: &Connection, opshash: &OpsHash, block_height: u64, first_block_height: u64, total_burn: u64) -> Result<ConsensusHash, db_error> {
+        let prev_consensus_hashes = BurnDB::<A, K>::get_prev_consensus_hashes(conn, block_height - 1, first_block_height)?;
+        Ok(ConsensusHash::from_ops(opshash, total_burn, &prev_consensus_hashes))
+    }
+
+    /// Find out how any burn tokens were destroyed in a given (canonical) block
+    fn find_block_burn_amount<'a>(tx: &mut Transaction<'a>, block_height: u64) -> Result<u64, db_error>
+    {
+        let user_burns = BurnDB::<A, K>::get_user_burns_by_block(tx, block_height)?;
+        let block_commits = BurnDB::<A, K>::get_block_commits_by_block(tx, block_height)?;
+        let mut burn_total = 0;
+
+        for i in 0..user_burns.len() {
+            if burn_total + user_burns[i].burn_fee < 0 {
```

Comment:
These checks won't find the overflow -- `burn_total` gets implicitly typed to an unsigned integer (because `.burn_fee` is unsigned, which means that when it overflows, it won't overflow to a negative number. I think compiler warnings should flag this line.

Reply:
Good catch!  Will fix